airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3StorageOperations.java

```diff
@@ -59,7 +60,30 @@ public String getBucketObjectPath(final String namespace, final String streamNam
             .replaceAll(Pattern.quote("${HOUR}"), String.format("%02d", writeDatetime.hourOfDay().get()))
             .replaceAll(Pattern.quote("${MINUTE}"), String.format("%02d", writeDatetime.minuteOfHour().get()))
             .replaceAll(Pattern.quote("${SECOND}"), String.format("%02d", writeDatetime.secondOfMinute().get()))
+            .replaceAll(Pattern.quote("${MILLISECOND}"), String.format("%02d", writeDatetime.millisOfSecond().get()))
+            .replaceAll(Pattern.quote("${EPOCH}"), String.format("%d", writeDatetime.getMillis()))
+            .replaceAll(Pattern.quote("${UUID}"), String.format("%s", UUID.randomUUID()))
             .replaceAll("/+", "/"));
+    if (customPathFormat.contains("${PART_ID}")) {
+      return objectPath.replaceAll(Pattern.quote("${PART_ID}"), String.format("%d", countObjectsInBucket(objectPath)));
+    } else {
+      return objectPath;
+    }
+  }
+
+  private int countObjectsInBucket(final String objectPath) {
```

Comment:
This method concerns me from the performance point of view, because it will be called for every stream and it has a while loop albeit the loop steps in object batches. Do we need sequential part IDs?

Reply:
I made a change to call `listObject` only once when this function is called.
If the batch is truncated (it would mean the bucket contains more than 1000 objects?), in that case I fall back to UUIDs...